src/scope.rs

```diff
@@ -1,292 +1,950 @@
-/// accessor callbacks; use FunctionCallbackScope and PropertyCallbackScope
-/// instead.
-///
-/// A CallbackScope can be created from the following inputs:
-///   - `Local<Context>`
-///   - `Local<Message>`
-///   - `Local<Object>`
-///   - `Local<Promise>`
-///   - `Local<SharedArrayBuffer>`
-///   - `&PromiseRejectMessage`
-pub struct CallbackScope<X = Contained> {
-  isolate: *mut Isolate,
-  phantom: PhantomData<X>,
-}
+  impl<'s> Deref for HandleScope<'s> {
+    type Target = ContextScope<'s>;
+    fn deref(&self) -> &Self::Target {
+      self.cast_ref()
+    }
+  }
 
-pub struct Contained;
-pub struct Escapable;
+  impl<'s> DerefMut for HandleScope<'s> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+      self.cast_mut()
+    }
+  }
 
-impl<'s> CallbackScope {
-  pub fn new<I>(input: I) -> Scope<'s, Self>
+  impl<'s, C> Drop for HandleScope<'s, C> {
+    fn drop(&mut self) {
+      data::ScopeData::get_mut(self).notify_scope_dropped();
+    }
+  }
+
+  pub unsafe trait NewHandleScopeParam<'s> {
+    type NewScope: Scope;
+    fn get_isolate_mut(&mut self) -> &mut Isolate;
+  }
+
+  unsafe impl<'s> NewHandleScopeParam<'s> for OwnedIsolate {
+    type NewScope = HandleScope<'s, ()>;
+    fn get_isolate_mut(&mut self) -> &mut Isolate {
+      &mut **self
+    }
+  }
+
+  unsafe impl<'s, 'p: 's, P> NewHandleScopeParam<'s> for ContextScope<'p, P>
   where
-    Scope<'s, Self>: From<I>,
+    P: Scope + NewHandleScopeParam<'s>,
   {
-    Scope::from(input)
+    type NewScope = <P as NewHandleScopeParam<'s>>::NewScope;
+    fn get_isolate_mut(&mut self) -> &mut Isolate {
+      data::ScopeData::get_mut(self).get_isolate_mut()
+    }
+  }
+
+  unsafe impl<'s, 'p: 's, C> NewHandleScopeParam<'s> for HandleScope<'p, C> {
+    type NewScope = HandleScope<'s, C>;
+    fn get_isolate_mut(&mut self) -> &mut Isolate {
+      data::ScopeData::get_mut(self).get_isolate_mut()
+    }
+  }
+
+  unsafe impl<'s, 'p: 's, 'e: 'p, C> NewHandleScopeParam<'s>
+    for EscapableHandleScope<'p, 'e, C>
+  {
+    type NewScope = EscapableHandleScope<'s, 'e, C>;
+    fn get_isolate_mut(&mut self) -> &mut Isolate {
+      data::ScopeData::get_mut(self).get_isolate_mut()
+    }
+  }
+
+  unsafe impl<'s, 'p: 's> NewHandleScopeParam<'s> for CallbackScope<'p> {
+    type NewScope = HandleScope<'s>;
+    fn get_isolate_mut(&mut self) -> &mut Isolate {
+      data::ScopeData::get_mut(self).get_isolate_mut()
+    }
+  }
+
+  /// A HandleScope which first allocates a handle in the current scope
+  /// which will be later filled with the escape value.
+  #[derive(Debug)]
+  pub struct EscapableHandleScope<'s, 'e: 's, C = Context> {
+    data: NonNull<data::ScopeData>,
+    _phantom:
+      PhantomData<(&'s mut raw::HandleScope, &'e raw::EscapeSlot, &'s C)>,
+  }
+
+  impl<'s, 'e: 's> EscapableHandleScope<'s, 'e> {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new<P>(param: &'s mut P) -> P::NewScope
```

Comment:
> Where is `NewScope` used?

I guess here... TBH I'm a little confused about what's happening here. Maybe you can update[ the "Design of Scopes" text](https://github.com/denoland/rusty_v8/blob/acdcede4d67ad4fab6d969f71eb3f99b5b4a13cc/src/lib.rs#L30-L65)?

Reply:
The answer to your question is probably somewhere here: https://github.com/denoland/rusty_v8/blob/375895dd55b5da4acda7a7caa9d1ffa710f615b7/src/scope.rs#L432-L444